File: src/src/com/makingiants/android/banjotuner/SoundPlayer.java
package com.makingiants.android.banjotuner;

import java.io.IOException;

import android.content.Context;
import android.content.res.AssetFileDescriptor;
import android.media.AudioManager;
import android.media.MediaPlayer;
import android.media.MediaPlayer.OnCompletionListener;
import android.media.MediaPlayer.OnPreparedListener;

/**
 * Sound player for android
 */
public class SoundPlayer implements OnPreparedListener, OnCompletionListener {

	// ****************************************************************
	// Attributes
	// ****************************************************************

	private final MediaPlayer mediaPlayer;
	private Context context;

	// ****************************************************************
	// Constructor
	// ****************************************************************

	public SoundPlayer(final Context context) {

		this.context = context;

		mediaPlayer = new MediaPlayer();
		mediaPlayer.setAudioStreamType(AudioManager.STREAM_MUSIC);
		mediaPlayer.setVolume(1.0f, 1.0f);
		mediaPlayer.setOnPreparedListener(this);
		mediaPlayer.setOnCompletionListener(this);
	}

	// ****************************************************************
	// Sound play methods
	// ****************************************************************

	/**
	 * Play the sound in path synchronized (only one sound at time)
	 * 
	 * @param path
	 *            file to play
	 * @throws IOException
	 */
	public void playSynchronized(final String path) throws IOException,
			InterruptedException {

		final AssetFileDescriptor afd = context.getAssets().openFd(path);

		synchronized (mediaPlayer) {

			mediaPlayer.setDataSource(afd.getFileDescriptor(),
					afd.getStartOffset(), afd.getLength());

			mediaPlayer.prepare();
			mediaPlayer.start();

			// Make the other synchronized calls to wait until the file
			// is
			// played
			Thread.sleep(mediaPlayer.getDuration());

			mediaPlayer.stop();
			mediaPlayer.reset();

		}

		afd.close();
	}

	/**
	 * Play the sound in path and if there are any sound playing it will stop
	 * and play the new one if there are no sounds playing now, else stop the
	 * sound
	 * 
	 * @param path
	 *            file to play
	 * @throws IOException
	 */
	public void play(final String path) throws IOException,
			InterruptedException {

		if (mediaPlayer.isPlaying()) {
			mediaPlayer.reset();
		}

		final AssetFileDescriptor afd = context.getAssets().openFd(path);

		mediaPlayer.setDataSource(afd.getFileDescriptor(),
				afd.getStartOffset(), afd.getLength());

		mediaPlayer.prepareAsync();

		afd.close();

	}

	// ****************************************************************
	// OnPreparedListener, OnCompletionListener implements
	// ****************************************************************

	public void onPrepared(final MediaPlayer player) {
		mediaPlayer.start();
	}

	public void onCompletion(final MediaPlayer arg0) {
		mediaPlayer.reset();
	}

}

File: app/src/main/java/com/makingiants/android/banjotuner/SoundPlayer.kt
package com.makingiants.android.banjotuner

import android.content.Context
import android.media.AudioManager
import android.media.MediaPlayer
import android.media.MediaPlayer.OnCompletionListener
import android.media.MediaPlayer.OnPreparedListener

/**
 * Sound player for android
 */
class SoundPlayer(private val context: Context) : OnPreparedListener, OnCompletionListener {
  private var mediaPlayer: MediaPlayer? = null
  private val audioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
  private val soundsPath = "b_sounds"
  private val sounds = arrayOf("1.mp3", "2.mp3", "3.mp3", "4.mp3")

  /**
   * Play the sound in path and if there are any sound playing it will stop
   * and play the new one if there are no sounds playing now, else stop the
   * sound

   * @param index file to play
   * *
   * @throws java.io.IOException
   */
  fun playWithLoop(index: Int) {
    if (mediaPlayer?.isPlaying == true) {
      stop()
    }

    val afd = context.assets.openFd("$soundsPath/${sounds[index]}")

    mediaPlayer = MediaPlayer().apply {
      setAudioStreamType(AudioManager.STREAM_MUSIC)
      setVolume(1.0f, 1.0f)
      setOnPreparedListener(this@SoundPlayer)
      setOnCompletionListener(this@SoundPlayer)

      setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)

      prepareAsync()
      isLooping = true
    }

    afd.close()
  }

  /**
   * Reset the player with a mute to avoid
   * a cut sound on reset.
   */
  fun stop() {
    audioManager.setStreamMute(AudioManager.STREAM_MUSIC, true)

    mediaPlayer?.apply {
      stop()
      reset()
      release()
    }

    mediaPlayer = null
    audioManager.setStreamMute(AudioManager.STREAM_MUSIC, false)
  }

  //<editor-fold desc="OnPreparedListener, OnCompletionListener implements">

  override fun onPrepared(player: MediaPlayer) {
    mediaPlayer?.start()
  }

  override fun onCompletion(arg0: MediaPlayer) {
    //        mediaPlayer.reset();
  }
  //</editor-fold>
}
